File: frontend/src/stores/graphStore/modules/v2/shared/embeddedConstraints.ts
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Copyright 2026 Precis Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @file embeddedConstraints.ts
 * @description V2 内嵌约束物化模块
 *
 * 将 schema.yaml 中内嵌的 constraints 字段物化为画布上的 Constraint 节点和边。
 * 通过 NodeDataBuilder 统一构建节点数据，确保与独立约束导入路径一致。
 *
 * 功能概述：
 * - materializeV2EmbeddedConstraints: 主入口，遍历 schema 的 embeddedConstraints
 * - 按约束类型通过 buildNodeData 创建对应的 Constraint 节点
 * - 自动建立 Constraint 节点到 Schema 列的边
 * - 去重检查：避免创建重复的约束节点
 *
 * 架构设计：
 * - 纯函数设计，接收 schemaNode + 工具函数作为参数
 * - 名称字段（column / from_column 等）严格解析：顶层裸名 / 嵌套「父.子」全限定路径，
 *   不接受列 ID、不做递归裸名猜测（columnRefResolver，与后端 embedded_constraints 同一约定）；
 *   解析不到视为配置错误，跳过该约束（后端同口径报错丢弃）
 * - ID 字段（then_column_id / if_column_id / column_id）按列 ID 直查
 * - 嵌套子列约束的连线挂到顶层祖先列的 handle（rootColumnId），节点展示/持久化用全限定名
 * - 使用 hasNode / addNode / addConstraintEdge 回调与外部状态交互
 */

import type { CustomNode } from '@/types/graph'
import {
  getConstraintKindByV2Type,
  getConstraintNodeTypeByV2Type,
} from '@/services/constraints/validationRegistry'
import type { BuildInput } from '@/services/constraints/nodeDataBuilder'
import { buildNodeData } from '@/services/constraints/nodeDataBuilder'
import {
  buildColumnRefResolver,
  type ColumnRefNode,
  type ResolvedColumnRef,
} from '@/services/constraints/columnRefResolver'
import { logger } from '@/core/utils/logger'
interface EmbeddedConstraintItem {
  id?: string | number
  type?: string
  column?: string
  description?: string
  // ForeignKey 内嵌约束的表/列名引用（宿主 schema 即 from_table，与 schemaBuilder 导出格式一致）
  from_table?: string
  from_column?: string
  to_table?: string
  to_column?: string
  params?: {
    allowed_values?: unknown[]
    min?: unknown
    max?: unknown
    expression?: unknown
    name?: unknown
    charset_mode?: unknown
    allowed_chars?: unknown
    disallowed_chars?: unknown
    logic_mode?: unknown
    start_date?: unknown
    end_date?: unknown
    date_format?: unknown
    [key: string]: unknown
  }
  // Conditional 内嵌约束的 refs 字段
  refs?: {
    if_conditions?: unknown[]
    if_logic?: string
    then_column_id?: string
    [key: string]: unknown
  }
}

export function materializeV2EmbeddedConstraints(params: {
  schemaNode: CustomNode
  schemaTableName: string
  embeddedConstraints: EmbeddedConstraintItem[]
  /** 宿主 schema 的列树（V2 文件/AI spec/JsonSchema 节点列），用于精确解析列引用 */
  columnTree: ColumnRefNode[]
  hasNode: (id: string) => boolean
  addNode: (node: CustomNode) => void
  addConstraintEdge: (tableId: string, constraintId: string, columnId: string) => void
}) {
  const {
    schemaNode,
    schemaTableName,
    embeddedConstraints,
    columnTree,
    hasNode,
    addNode,
    addConstraintEdge,
  } = params

  // 名称字段严格解析：顶层裸名 / 嵌套「父.子」全限定路径（与后端 embedded_constraints 同约定）
  const resolver = buildColumnRefResolver(columnTree)
  /** 列 ID -> 全限定显示名（未知 ID 返回空串） */
  const columnNameById = (columnId: string): string => resolver.byId(columnId)?.columnName ?? ''
  /** 列 ID -> 顶层祖先列 id（连线 handle 用；顶层列即自身） */
  const rootColumnIdOf = (columnId: string): string =>
    resolver.byId(columnId)?.rootColumnId ?? columnId
  /** if/then 列引用解析：ID 字段按列 ID 直查（不接受列名） */
  const resolveRefColumnId = (value: string): string => {
    if (!value) return ''
    return resolver.byId(value)?.columnId ?? ''
  }

  embeddedConstraints.forEach((item: EmbeddedConstraintItem, idx: number) => {
    if (!item?.id) return

    const rawId = String(item.id)
    const id = rawId.startsWith(`${schemaNode.id}_`) ? rawId : `${schemaNode.id}_${rawId}`

    if (hasNode(id)) return

    const nodeType = getConstraintNodeTypeByV2Type(item.type ?? '') ?? 'constraint'
    const basePos = { x: schemaNode.position.x + 420, y: schemaNode.position.y + idx * 160 }
    const kind = getConstraintKindByV2Type(item.type ?? '')

    // 名称字段严格解析：顶层裸名 / 嵌套「父.子」全限定路径，不接受列 ID。
    // Conditional 除外：其 column 字段承载 THEN 列 ID（保存契约），属 ID 语义，按 ID 直查
    const isConditional = item.type === 'Conditional'
    const colName = !isConditional && item.column ? String(item.column) : ''
    const resolvedColumn: ResolvedColumnRef | undefined = colName
      ? resolver.resolve(colName)
      : undefined
    if (colName && !resolvedColumn) {
      // 与后端 embedded_constraints 同口径：解析不到视为配置错误，跳过该约束
      logger.warn(
        `[embeddedConstraints] 内嵌约束 ${id} 的列引用 "${colName}" 在 schema ${schemaNode.id} 中无法解析，已跳过（顶层列用裸名，嵌套子列用「父.子」路径）`
      )
      return
    }

    // 构建 BuildInput
    let buildInput: BuildInput

    if (item.type === 'Conditional') {
      // Conditional 内嵌约束 — 解析 IF 条件和 THEN 列
      // 优先从 params 读取（与 schemaBuilder / embeddedConstraintBuilder 导出格式一致），兼容旧版 refs
      const itemParams = (item.params || {}) as Record<string, unknown>
      const itemRefs = (item.refs || itemParams) as Record<string, unknown>
      const ifLogic = String(itemRefs.if_logic || itemParams.if_logic || 'and')
      // THEN 列：then_column_id 即列 ID，按 ID 直查；旧格式（无 then_column_id）回退 base.column（同样按 ID 解析）
      const thenRaw = String(
        itemRefs.then_column_id || itemParams.then_column_id || item.column || ''
      )
      const thenColId = resolveRefColumnId(thenRaw)

      const rawConditions = Array.isArray(itemRefs.if_conditions || itemParams.if_conditions)
        ? ((itemRefs.if_conditions || itemParams.if_conditions) as unknown[])
        : []
      const ifConditions = rawConditions.map((cond) => {
        const r = cond as Record<string, unknown>
        // if_column_id 即列 ID，直通（UUID 不再误当列名查 colNameToId）
        const ifColId = resolveRefColumnId(String(r?.if_column_id || ''))
        return {
          operator: String(r?.operator ?? ''),
          value: r?.value,
          values: r?.values as unknown[] | undefined,
          columnId: ifColId,
          columnName: columnNameById(ifColId),
        }
      })

      buildInput = {
        mode: 'embedded',
        configName: item.description || id,
        schemaNodeId: schemaNode.id,
        tableName: schemaTableName,
        nodeId: id,
        nodeType,
        embedded: true,
        ifConditions,
        ifLogic,
        thenRef: thenColId
          ? {
              nodeId: schemaNode.id,
              columnId: thenColId,
              columnName: columnNameById(thenColId),
            }
          : undefined,
        thenConditionConfig: itemParams.then_condition,
        // params 透传：skip_if 开关由 builder 读取（input.params.skip_if），保证 roundtrip 不复位
        params: item.params as Record<string, unknown> | undefined,
      }
    } else if (item.type === 'ForeignKey') {
      // ForeignKey 内嵌约束 — from_table/from_column/to_table/to_column 均为名（列名简化写法）。
      // 宿主 schema 即 from_table（与后端 embedded_constraints 一致，refs.from_table_id = schema.id）；
      // to_table 为跨表引用，物化入口无其他 schema 节点的访问能力——目标按名保留
      // （targetRef 置空、不建 FK 展示边），需要全量恢复时走独立导入路径（import/constraint.ts）
      const fromColName = item.from_column ? String(item.from_column) : ''
      const fromResolved = fromColName ? resolver.resolve(fromColName) : undefined
      if (fromColName && !fromResolved) {
        // 名称字段严格解析：解析不到视为配置错误，跳过该约束（与后端同口径）
        logger.warn(
          `[embeddedConstraints] 内嵌 FK 约束 ${id} 的列引用 "${fromColName}" 在 schema ${schemaNode.id} 中无法解析，已跳过`
        )
        return
      }
      const fromColId = fromResolved?.columnId ?? ''
      const toTableName = item.to_table ? String(item.to_table) : ''
      const toColName = item.to_column ? String(item.to_column) : ''

      buildInput = {
        mode: 'embedded',
        configName: item.description || id,
        schemaNodeId: schemaNode.id,
        tableName: schemaTableName,
        nodeId: id,
        nodeType,
        embedded: true,
        fkRefs: {
          source: {
            nodeId: schemaNode.id,
            columnId: fromColId,
            columnName: fromResolved?.columnName ?? fromColName,
          },
          target: { nodeId: '', columnId: '', columnName: toColName },
        },
        // builder 从 refs.to_table_name 取目标表显示名
        refs: toTableName ? { to_table_name: toTableName } : undefined,
        // params 透传：allow_null 开关由 builder 读取（input.params.allow_null），保证 roundtrip 不复位
        params: item.params as Record<string, unknown> | undefined,
      }
    } else {
      // 通用单列约束
      buildInput = {
        mode: 'embedded',
        configName: item.description || id,
        schemaNodeId: schemaNode.id,
        tableName: schemaTableName,
        nodeId: id,
        nodeType,
        embedded: true,
        columnRef: resolvedColumn
          ? {
              nodeId: schemaNode.id,
              columnId: resolvedColumn.columnId,
              columnName: resolvedColumn.columnName,
            }
          : undefined,
        params: item.params as Record<string, unknown> | undefined,
      }
    }

    // 使用 NodeDataBuilder 构建节点数据
    const result = kind
      ? buildNodeData(kind, buildInput)
      : {
          // 未知类型的降级
          nodeData: {
            embedded: true,
            configName: item.description || id,
            saveState: 'saved',
          } as Record<string, unknown>,
          edgeDescriptors: [] as Array<{
            kind: 'constraint'
            sourceNodeId: string
            targetNodeId: string
            columnId: string
          }>,
        }

    addNode({
      id,
      type: nodeType,
      position: basePos,
      // builder 产出宽松 Record<string, unknown>，与 CustomNodeData 联合无足够重叠，
      // 单断言编译不过（TS2352），此处为注册表 builder 出口的标准接收方式（同 import/constraint.ts）
      data: result.nodeData as unknown as CustomNode['data'],
    })

    // 创建边（嵌套子列约束挂到顶层祖先列的 handle）
    for (const desc of result.edgeDescriptors) {
      if (desc.kind === 'constraint' || desc.kind === 'if') {
        addConstraintEdge(desc.sourceNodeId, desc.targetNodeId, rootColumnIdOf(desc.columnId))
      }
    }
  })
}

File: backend/app/shared/core/project/loader/loader_parts/embedded_constraints.py
# SPDX-License-Identifier: Apache-2.0
#
# Copyright 2026 Precis Team
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
@fileoverview 内嵌约束收集模块

功能概述:
- 从 Schema 文件中收集内嵌的约束定义
- 将 Schema 文件中的 inline 约束转换为独立的 ConstraintFile

架构设计:
- 扫描所有 schema 的 constraints 字段
- 转换约束引用: 全限定列名 -> column id (规范化)

列引用约定(精确匹配,不做递归裸名猜测):
- 顶层列用裸名: `email` 仅匹配顶层列 email
- 嵌套子列用「父.子」点分路径: `customer.email` 仅匹配 customer 下的 email
- 未命中的引用原样保留(兼容直接写列 ID 的引用,真正不存在时由 factory 阶段报错)

输入示例:
    schema_files = {
        "users": TableSchemaFile(
            id="users",
            name="用户表",
            columns=[
                Column(id="col_1", name="id", type="integer"),
                Column(id="col_2", name="username", type="string"),
                Column(id="col_3", name="email", type="string")
            ],
            constraints=[
                ConstraintItem(
                    id="not_null_username",
                    type="NotNull",
                    column="username",  # 列名 (需要转换为 id)
                    enabled=True
                ),
                ConstraintItem(
                    id="fk_orders",
                    type="ForeignKey",
                    from_column="user_id",
                    to_table="users",
                    to_column="id",
                    enabled=True
                )
            ]
        )
    }

输出示例:
    constraint_files = {
        "users_not_null_username": ConstraintFile(
            version=2,
            id="users_not_null_username",
            type="NotNull",
            enabled=True,
            refs={"table_id": "users", "column_id": "col_2"},
            params={}
        ),
        "users_fk_orders": ConstraintFile(
            version=2,
            id="users_fk_orders",
            type="ForeignKey",
            enabled=True,
            refs={
                "from_table_id": "users",
                "from_column_id": "user_id",  # 保持原名（未找到对应 column）
                "to_table_id": "users",
                "to_column_id": "id"
            },
            params={}
        )
    }
"""

from __future__ import annotations

from typing import Any, cast

from app.shared.core.project.constraint.registry import normalize_constraint_type
from app.shared.core.project.constraint.types import ConstraintFile
from app.shared.core.project.loader.types import LoadingError
from app.shared.core.project.schema.types import TableSchemaFile
from app.shared.core.project.schema.types_parts.column_utils import build_qualified_name_to_id_map


def collect_constraints_from_schemas(
    schema_files: dict[str, TableSchemaFile],
    loading_errors: list[LoadingError] | None = None,
    schema_paths: dict[str, str] | None = None,
) -> dict[str, ConstraintFile]:
    """@methoddesc 从 Schema 文件中收集内嵌约束

    遍历所有 Schema，提取其中的 constraints 字段，
    转换为独立的 ConstraintFile 对象。

    核心转换逻辑:
    1. 约束 ID: "{table_id}_{constraint_id}" 格式，确保全局唯一
    2. 列引用: 全限定列名 -> column id (顶层裸名/嵌套点分路径,精确匹配)
    3. 外键特殊处理: from_column/to_column 也需要转换

    列引用规则 (严格模式,无任何兜底):
        `column` / `columns` / `from_column` / `to_column` 只接受顶层列裸名
        或嵌套「父.子」全限定路径;解析不到视为配置错误,记入 loading_errors
        并丢弃该约束(不猜测、不当作列 ID 直通)。列 ID 只出现在 column_id /
        column_ids 等 ID 字段(独立约束文件、Conditional 的 then_column_id)。

    输入示例:
        schema_files = {
            "users": TableSchemaFile(
                id="users",
                constraints=[
                    ConstraintItem(
                        id="not_null_email",
                        type="NotNull",
                        column="email",  # 列名
                        enabled=True,
                        params={}
                    )
                ]
            )
        }

    输出示例:
        constraint_files = {
            "users_not_null_email": ConstraintFile(
                version=2,
                id="users_not_null_email",
                type="NotNull",
                enabled=True,
                refs={"table_id": "users", "column_id": "col_3"},  # email 的 column id
                params={}
            )
        }

    原理说明:
        - normalize_constraint_type() 确保约束类型名称标准化 (如 "not null" -> "NotNull")
        - 列引用到列 ID 的转换: 按「顶层裸名 / 嵌套「父.子」全限定路径」精确匹配,
          不做递归裸名猜测;未命中即报错并丢弃该约束
        - ConstraintFile.version 固定为 2 (当前版本)
    """
    # 用于存储转换后的约束文件对象，键为约束全局唯一 ID
    constraint_files: dict[str, ConstraintFile] = {}
    # 遍历所有已加载的 Schema 文件
    for schema in schema_files.values():
        # 如果当前 Schema 没有定义任何内嵌约束，则直接跳过
        if not schema.constraints:
            continue

        # 全限定列名 -> column_id 解析表:顶层列用裸名,嵌套列用「父.子」点分路径,
        # 精确匹配不做递归裸名猜测(避免同名子列误绑定)
        ref_to_id = build_qualified_name_to_id_map(schema.columns)

        def _report_bad_ref(constraint_id: str, ref: str) -> None:
            """列引用无法解析:结构化报错(调用方传入 loading_errors 时)。"""
            if loading_errors is None:
                return
            loading_errors.append(
                LoadingError(
                    error_type="EmbeddedColumnRefError",
                    file_path=(schema_paths or {}).get(schema.id, ""),
                    ref_id=constraint_id,
                    severity="blocker",
                    title="内嵌约束的列引用无法解析",
                    message=f"约束 '{constraint_id}' 引用的列 '{ref}' 在表 '{schema.id}' 中不存在",
                    suggestion=(
                        "顶层列用裸名（如 email），嵌套子列用「父.子」全限定路径（如 customer.email）；"
                        f"可用列: {sorted(ref_to_id)}"
                    ),
                )
            )

        # 遍历该 Schema 中定义的每一条内嵌约束
        for constraint_item in schema.constraints:
            # 将约束类型名称标准化（如 "not null" 转换为 "NotNull"）
            constraint_type = normalize_constraint_type(constraint_item.type)

            # 初始化 refs 字典，至少包含当前表 ID
            refs: dict[str, Any] = {"table_id": schema.id}
            # 约束是否因列引用无法解析而被丢弃
            dropped = False
            # 通用名称字段解析（Conditional 除外：其 column 字段承载 THEN 列 ID，属 ID 语义）
            if constraint_type != "Conditional" and constraint_item.column:
                col_id = ref_to_id.get(constraint_item.column)
                if col_id is None:
                    _report_bad_ref(f"{schema.id}_{constraint_item.id}", constraint_item.column)
                    dropped = True
                else:
                    refs["column_id"] = col_id
            elif constraint_type != "Conditional" and constraint_item.columns:
                column_ids = []
                for col_name in constraint_item.columns:
                    col_id = ref_to_id.get(col_name)
                    if col_id is None:
                        _report_bad_ref(f"{schema.id}_{constraint_item.id}", col_name)
                        dropped = True
                        break
                    column_ids.append(col_id)
                if not dropped:
                    refs["column_ids"] = column_ids

            # 外键约束需要特殊处理：涉及源列和目标列的映射
            if not dropped and constraint_type == "ForeignKey":
                # 源列引用（顶层列名或嵌套全限定路径）解析为列 ID（未命中报错丢弃）
                from_col_id = None
                if constraint_item.from_column:
                    from_col_id = ref_to_id.get(constraint_item.from_column)
                    if from_col_id is None:
                        _report_bad_ref(f"{schema.id}_{constraint_item.id}", constraint_item.from_column)
                        dropped = True

                # 对目标列也执行列引用 → 列 ID 的转换（在目标 schema 中按同一约定严格解析）
                to_col_id = constraint_item.to_column
                if not dropped and constraint_item.to_table and constraint_item.to_column:
                    to_schema = schema_files.get(constraint_item.to_table)
                    if to_schema:
                        to_ref_to_id = build_qualified_name_to_id_map(to_schema.columns)
                        to_col_id = to_ref_to_id.get(constraint_item.to_column)
                        if to_col_id is None:
                            _report_bad_ref(f"{schema.id}_{constraint_item.id}", constraint_item.to_column)
                            dropped = True

                if not dropped:
                    # 外键的 refs 结构包含源表/列和目标表/列
                    refs = {
                        "from_table_id": schema.id,
                        "from_column_id": from_col_id,
                        "to_table_id": constraint_item.to_table,
                        "to_column_id": to_col_id,
                    }

            if dropped:
                continue

            # Conditional 约束：将 params 中的字段提取到 refs（前端 embedded 约束无 refs 字段）
            # 提取后从 params 中移除，避免下游同时看到 refs 和 params 中的重复字段
            if constraint_type == "Conditional":
                params = dict(constraint_item.params or {})
                refs["if_logic"] = params.pop("if_logic", "and")
                refs["if_conditions"] = params.pop("if_conditions", []) or []
                refs["then_column_id"] = params.pop("then_column_id", None)
                params.pop("table_id", None)
            else:
                params = constraint_item.params or {}

            # 构建 ConstraintFile 对象，将内嵌约束转换为独立约束文件格式
            cf = ConstraintFile.model_construct(
                version=2,
                id=f"{schema.id}_{constraint_item.id}",  # 使用 "{table_id}_{constraint_id}" 确保全局唯一
                type=cast(Any, constraint_type),
                enabled=constraint_item.enabled,
                refs=refs,
                params=params,
            )
            # 以全局唯一 ID 为键存入字典
            constraint_files[cf.id] = cf

    # 返回所有收集到的约束文件对象
    return constraint_files
